File: depth_first_search/01_dfs_flood_fill/dfs_flood_fill.py
import numpy as np


class Solution:
    def flood_fill_util(self, image: np.ndarray,
                        sr: int, sc: int,
                        new_color: int, old_color: int) -> None:
        if (sr < 0 or sr >= image.shape[0]) or (sc < 0 or sc >= image.shape[1]) or (image[sr, sc] != old_color or image[sr, sc] == new_color):
            return

        image[sr, sc] = new_color

        self.flood_fill_util(image, sr + 1, sc, new_color, old_color)
        self.flood_fill_util(image, sr - 1, sc, new_color, old_color)
        self.flood_fill_util(image, sr, sc + 1, new_color, old_color)
        self.flood_fill_util(image, sr, sc - 1, new_color, old_color)

    def flood_fill(self, image: np.ndarray,
                   sr: int, sc: int,
                   new_color: int) -> np.ndarray:
        if not self._check_constraints(image, sr, sc, new_color):
            raise ValueError('Constraints violated')

        old_color = image[sr, sc]
        self.flood_fill_util(image, sr, sc, new_color, old_color)

        return image

    def _check_constraints(self, image: np.ndarray,
                           sr: int, sc: int,
                           new_color: int) -> bool:
        m = image.shape[0]
        if m == 0 or m > 50:
            return False

        for row in image:
            n = row.shape[0]
            if n == 0 or n > 50:
                return False

            for color in row:
                if color < 0 or color >= 2**16:
                    return False

        if new_color < 0 or new_color >= 2**16:
            return False

        if (sr < 0 or sr >= m) or (sc < 0 or sc >= n):
            return False

        return True

File: depth_first_search/01_dfs_flood_fill/dfs_flood_fill.cpp
#include <vector>
#include <cassert>
#include <cmath>

using namespace std;


class Solution {
public:
    void floodFillUtil(vector<vector<int>>& image, int sr, int sc, int newColor, int oldColor) {
        // Check for out-of-bounds or if the color is already the new color or not the old color
        if (sr < 0 || sr >= image.size() || sc < 0 || sc >= image[sr].size() || 
            image[sr][sc] != oldColor || image[sr][sc] == newColor) {
            return;
        }

        // Fill the color
        image[sr][sc] = newColor;

        // Recursively call function for the surrounding pixels (up, down, left, right)
        floodFillUtil(image, sr + 1, sc, newColor, oldColor);
        floodFillUtil(image, sr - 1, sc, newColor, oldColor);
        floodFillUtil(image, sr, sc + 1, newColor, oldColor);
        floodFillUtil(image, sr, sc - 1, newColor, oldColor);
    }

    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        assert(checkConstraints(image, sr, sc, newColor));

        int oldColor = image[sr][sc];
        floodFillUtil(image, sr, sc, newColor, oldColor);
        return image;
    }

private:
    // Helper function to check the constraints
    bool checkConstraints(const vector<vector<int>>& image, int sr, int sc, int newColor) {
        int m = image.size(); 
        if (m == 0 || m > 50) return false;
        for (const auto& row : image) {
            int n = row.size();
            if (n == 0 || n > 50) return false;
            for (int color : row) {
                if (color < 0 || color >= pow(2, 16)) return false;
            }
        }
        if (newColor < 0 || newColor >= pow(2, 16)) return false;
        if (sr < 0 || sr >= m || sc < 0 || sc >= image[0].size()) return false;
        
        return true;
    }
};
